core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -64,672 +83,1650 @@
-            }
-            else {
-                // all sources are remote
-                NodePartitionMap nodePartitionMap = partitioningCache.apply(plan.getFragment().getPartitioning());
-                List<InternalNode> partitionToNode = nodePartitionMap.getPartitionToNode();
-                // todo this should asynchronously wait a standard timeout period before failing
-                checkCondition(!partitionToNode.isEmpty(), NO_NODES_AVAILABLE, "No worker nodes available");
-                stageSchedulers.put(stageId, new FixedCountScheduler(stage, partitionToNode));
-                bucketToPartition = Optional.of(nodePartitionMap.getBucketToPartition());
             }
-            childStages = createChildStages.apply(bucketToPartition);
-        }
 
-        stage.addStateChangeListener(newState -> {
-            if (newState == FLUSHING || newState.isDone()) {
-                childStages.forEach(SqlStageExecution::cancel);
+            if (state == DistributedStagesSchedulerState.FAILED) {
+                StageFailureInfo stageFailureInfo = distributedStagesScheduler.getFailureCause()
+                        .orElseGet(() -> new StageFailureInfo(toFailure(new VerifyException("distributedStagesScheduler failed but failure cause is not present")), Optional.empty()));
+                ErrorCode errorCode = stageFailureInfo.getFailureInfo().getErrorCode();
+                if (shouldRetry(errorCode)) {
+                    long delayInMillis = min(retryInitialDelay.toMillis() * ((long) pow(2, currentAttempt.get())), retryMaxDelay.toMillis());
+                    currentAttempt.incrementAndGet();
+                    scheduleRetryWithDelay(delayInMillis);
+                }
+                else {
+                    stageManager.getDistributedStagesInTopologicalOrder().forEach(stage -> {
+                        if (stageFailureInfo.getFailedStageId().isPresent() && stageFailureInfo.getFailedStageId().get().equals(stage.getStageId())) {
+                            stage.fail(stageFailureInfo.getFailureInfo().toException());
+                        }
+                        else {
+                            stage.abort();
+                        }
+                    });
+                    queryStateMachine.transitionToFailed(stageFailureInfo.getFailureInfo().toException());
+                }
             }
         });
+        return Optional.of(distributedStagesScheduler);
+    }
 
-        stageLinkages.put(stageId, new StageLinkage(plan.getFragment().getId(), parent, childStages));
+    private boolean shouldRetry(ErrorCode errorCode)
+    {
+        return retryPolicy == RetryPolicy.QUERY && currentAttempt.get() < maxRetryAttempts && (errorCode == null || errorCode.getType() != USER_ERROR);
+    }
 
-        return stages.build();
+    private void scheduleRetryWithDelay(long delayInMillis)
+    {
+        try {
+            schedulerExecutor.schedule(this::scheduleRetry, delayInMillis, MILLISECONDS);
```

Comment:
Assuming the query state remains RUNNING during this delay period, is there some easy way to tell from the web ui that the query is waiting for a retry and not just stuck due to some issue ?

Reply:
During the delay between attempts all stages will be in a "PENDING" state. This indicates that no active tasks are currently running in a stage. We have a separate work item to update UI to make it easier to identify failures and retries.